### ISSUE TITLE
Add ability to enhance a structural type

### ISSUE THREAD

**Issue by DPUkyle** [gosu-lang/gosu-lang#71]
Here's a feature which would be cool to have: enhancing a structure.

Here are some sample classes to help illustrate what I mean:

Base type; pretend that we don't control its content:

``` gosu
structure BaseType {
  property get Foo() : String
  property get Bar() : String
}
```

Additional property we use to "decorate" the base type:

``` gosu
structure MyThing extends BaseType {
  property set SomeValue(value : Boolean)
  property get SomeValue() : Boolean
}
```

Now I make a type which is structurally assignable to BaseType

``` gosu
class Whatever {
  var foo : String as Foo
  var bar : String as Bar
}
```

Now write a program to work with BaseType, but requires an ugly cast to MyThing plus null-safety checks:

``` gosu
var x : BaseType = new Whatever()
if((x as MyThing).SomeValue == null or (x as MyThing).SomeValue == false) {
  ...
}
```

It would be cool if we could enhance BaseType to abstract the cast and null safety check, like this:

``` gosu
enhancement BaseTypeEnhancement : BaseType {
  property get SomeValue() : boolean {
    return (this as MyThing).SomeValue ?: false
  }
  property set SomeValue(value : boolean) {
    (this as MyThing).SomeValue = value
  }
}
```

Now the program would look like this:

``` gosu
var x : BaseType = new Whatever()
if(!x.SomeValue) { //mmm
  ...
}
```

**Comment by DPUkyle** [comment]
Here's a better example that's more readable.

Given the structure Foo:
```
structure Foo {
  property get Bar() : String
}
```
And an enhancement to it:
```
enhancement FooEnhancement : Foo {
  function PrintContentsOfBar() {
    print(this.Bar)
  }
}
```

I should be able to access the enhancement methods:
```
var x = new Foo() {
  var b: String as readonly Bar = "hi"
}

x.PrintContentsOfBar() // parsing and compilation errors, see notes below
```

There is some odd behavior:
* In Gosu Lab, the code executes successfully _from source_
* In Gosu Lab, the parser complains about `x.PrintContentsOfBar()`: `No function descriptor found for function, PrintContentsOfBar, on class, AnonymouS__1`
* Statically compiling the code results in a similar error `Error: (5, 1) No function descriptor found for function, PrintContentsOfBar, on class, AnonymouS__1`

So the runtime behavior seems to be OK, but there is a lingering issue with parsing.

**Comment by DPUkyle** [comment]
Fixed by 1c9b13c05736d637b5779a5735b91ec7b8ac51c3
